In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [2]:
df=pd.read_csv("cars.csv")
df

symboling normalized-losses         make fuel-type   body-style  \
0            3                 ?  alfa-romero       gas  convertible   
1            3                 ?  alfa-romero       gas  convertible   
2            1                 ?  alfa-romero       gas    hatchback   
3            2               164         audi       gas        sedan   
4            2               164         audi       gas        sedan   
..         ...               ...          ...       ...          ...   
200         -1                95        volvo       gas        sedan   
201         -1                95        volvo       gas        sedan   
202         -1                95        volvo       gas        sedan   
203         -1                95        volvo    diesel        sedan   
204         -1                95        volvo       gas        sedan   

    drive-wheels engine-location  width  height engine-type  engine-size  \
0            rwd           front   64.1    48.8        dohc          130   
1            rwd           front   64.1    48.8        dohc          130   
2            rwd           front   65.5    52.4        ohcv          152   
3            fwd           front   66.2    54.3         ohc          109   
4            4wd           front   66.4    54.3         ohc          136   
..           ...             ...    ...     ...         ...          ...   
200          rwd           front   68.9    55.5         ohc          141   
201          rwd           front   68.8    55.5         ohc          141   
202          rwd           front   68.9    55.5        ohcv          173   
203          rwd           front   68.9    55.5         ohc          145   
204          rwd           front   68.9    55.5         ohc          141   

    horsepower  city-mpg  highway-mpg  price  
0          111        21           27  13495  
1          111        21           27  16500  
2          154        19           26  16500  
3          102        24           30  13950  
4          115        18           22  17450  
..         ...       ...          ...    ...  
200        114        23           28  16845  
201        160        19           25  19045  
202        134        18           23  21485  
203        106        26           27  22470  
204        114        19           25  22625  

[205 rows x 15 columns]

In [3]:
df.head()


symboling normalized-losses         make fuel-type   body-style  \
0          3                 ?  alfa-romero       gas  convertible   
1          3                 ?  alfa-romero       gas  convertible   
2          1                 ?  alfa-romero       gas    hatchback   
3          2               164         audi       gas        sedan   
4          2               164         audi       gas        sedan   

  drive-wheels engine-location  width  height engine-type  engine-size  \
0          rwd           front   64.1    48.8        dohc          130   
1          rwd           front   64.1    48.8        dohc          130   
2          rwd           front   65.5    52.4        ohcv          152   
3          fwd           front   66.2    54.3         ohc          109   
4          4wd           front   66.4    54.3         ohc          136   

  horsepower  city-mpg  highway-mpg  price  
0        111        21           27  13495  
1        111        21           27  16500  
2        154        19           26  16500  
3        102        24           30  13950  
4        115        18           22  17450

Now,I can see ?,that means something is wrong. Solet's check info.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   body-style         205 non-null    object 
 5   drive-wheels       205 non-null    object 
 6   engine-location    205 non-null    object 
 7   width              205 non-null    float64
 8   height             205 non-null    float64
 9   engine-type        205 non-null    object 
 10  engine-size        205 non-null    int64  
 11  horsepower         205 non-null    object 
 12  city-mpg           205 non-null    int64  
 13  highway-mpg        205 non-null    int64  
 14  price              205 non-null    int64  
dtypes: float64(2), int64(5), object(8)
memory usage: 24.1+ KB


Now,there are 205 entries from range 0 to 204, and looks like there is no value missing. Check for isna()

In [5]:
df.isna().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
body-style           0
drive-wheels         0
engine-location      0
width                0
height               0
engine-type          0
engine-size          0
horsepower           0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

No missing value. But wehave seen missing values in normalized-loses.Now,what to do..?Missing values can be in many format like- NaN, ?, missing, NA, " ". By default info and isna detects NaN. Now,if u see normalized-loses is a numeric col. But check it's datatype,because ? is a object,that is why there might be a possibility  that there can be NaN.

Width and height are float64.

  engine-size        205 non-null    int64  
  horsepower         205 non-null    object 
  city-mpg           205 non-null    int64  
  highway-mpg        205 non-null    int64  
  price              205 non-null    int64 

  Horsepower has object datatype.  

In [6]:
df["normalized-losses"].value_counts()

?      41
161    11
91      8
150     7
134     6
128     6
104     6
85      5
94      5
65      5
102     5
74      5
168     5
103     5
95      5
106     4
93      4
118     4
148     4
122     4
83      3
125     3
154     3
115     3
137     3
101     3
119     2
87      2
89      2
192     2
197     2
158     2
81      2
188     2
194     2
153     2
129     2
108     2
110     2
164     2
145     2
113     2
256     1
107     1
90      1
231     1
142     1
121     1
78      1
98      1
186     1
77      1
Name: normalized-losses, dtype: int64

In [7]:
df["horsepower"].value_counts()

68     19
70     11
69     10
116     9
110     8
95      7
88      6
62      6
101     6
160     6
114     6
84      5
97      5
102     5
145     5
82      5
76      5
111     4
92      4
123     4
86      4
90      3
73      3
85      3
207     3
182     3
121     3
152     3
112     2
56      2
161     2
156     2
94      2
52      2
?       2
162     2
155     2
184     2
100     2
176     2
55      1
262     1
134     1
115     1
140     1
48      1
58      1
60      1
78      1
135     1
200     1
64      1
120     1
72      1
154     1
288     1
143     1
142     1
175     1
106     1
Name: horsepower, dtype: int64

41 ? in normalised-loses,and 2 in horsepower.

Now, fillna is applicable only on NaN,because by default pandas and sklearn can detect NaN. So we need to convert ? into NaN,also we need to change the datatype,becuse mean can be performed on numeric data and not on object data. And then use fillna.

In [8]:
#step1 -: replace the ? with NAN
df["normalized-losses"].replace("?", np.nan, inplace=True)
df["horsepower"].replace("?", np.nan, inplace=True)

#step2 -: change the datatype
df["normalized-losses"] = df["normalized-losses"].astype("float64") # now, we cannot write inplace for 
# permanent change, that's why assignment is done.
df["horsepower"] = df["horsepower"].astype("float64")

#step3 -: get the mean
nlmean =df["normalized-losses"].mean()
hpmean =df["horsepower"].mean()

#step4 -: use fillna
df["normalized-losses"].fillna(nlmean, inplace=True)
df["horsepower"].fillna(hpmean, inplace=True)

In [9]:
df.head()

symboling  normalized-losses         make fuel-type   body-style  \
0          3              122.0  alfa-romero       gas  convertible   
1          3              122.0  alfa-romero       gas  convertible   
2          1              122.0  alfa-romero       gas    hatchback   
3          2              164.0         audi       gas        sedan   
4          2              164.0         audi       gas        sedan   

  drive-wheels engine-location  width  height engine-type  engine-size  \
0          rwd           front   64.1    48.8        dohc          130   
1          rwd           front   64.1    48.8        dohc          130   
2          rwd           front   65.5    52.4        ohcv          152   
3          fwd           front   66.2    54.3         ohc          109   
4          4wd           front   66.4    54.3         ohc          136   

   horsepower  city-mpg  highway-mpg  price  
0       111.0        21           27  13495  
1       111.0        21           27  16500  
2       154.0        19           26  16500  
3       102.0        24           30  13950  
4       115.0        18           22  17450

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    float64
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   body-style         205 non-null    object 
 5   drive-wheels       205 non-null    object 
 6   engine-location    205 non-null    object 
 7   width              205 non-null    float64
 8   height             205 non-null    float64
 9   engine-type        205 non-null    object 
 10  engine-size        205 non-null    int64  
 11  horsepower         205 non-null    float64
 12  city-mpg           205 non-null    int64  
 13  highway-mpg        205 non-null    int64  
 14  price              205 non-null    int64  
dtypes: float64(4), int64(5), object(6)
memory usage: 24.1+ KB
